Unify TLP and RTO into Probe Timeout

janaiyengar opened the pull request:
Closes #2166, #2168, #1017.

This PR builds on #1974 and simplifies QUIC's timeout-based loss detection mechanisms. Unfortunately, this PR cuts through text in many places, so I'm summarizing the key changes here.  The compiled version of the draft, [here](https://quicwg.org/base-drafts/pto/draft-ietf-quic-recovery.html), might be easier to read than the diffs.

The high-order description is that TLP and RTO are combined in this PR into a single "Probe Timeout" (PTO). Consecutive PTOs are exponentially longer, just like RTO used to be.  The rationale for calling this a "Probe" Timeout is that loss detection in QUIC only happens when an ack is received, and this timer and timeout only cause probe packets to be sent out to elicit acks. This unification is made possible by the following constituent changes.

- (Issue #1017) There was general agreement in #1017 that adding a max_ack_delay should allow us to remove the minRTO. max_ack_delay has since been added to QUIC in a transport parameter, which allows us to include it in the PTO computation and remove minRTO/minTLP.

- Timer granularity is added for both crypto and PTO timeouts, ensuring that even when the computed network RTT might be too small and represented as 0, timeouts are set for at least 1 clock tick. This is in keeping with use of granularity in TCP (RFC 6298). A min granularity of 1ms is recommended.

- (Issue #2166) In line with the resolution above for removing min RTO, min TLP and min crypto timeout values are now removed.

- (Issue #2168) PTO validation has been removed.  PTOs cause congestion window collapse to min cwnd only after 3 consecutive PTOs to retain parity with current behavior, where an RTO collapses the window after 2 TLPs.

One side-effect to note is that a single timer will start exponentially backing off right from the first PTO, as against TLP and RTO where the TLP timers did not back off at all, and the RTO timer started exponentially backing off after the TLPs. I can restore this by adjusting the backoff expression in SetLossDetectionTimer(), but I'm not convinced this is necessary.

siyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -726,10 +700,9 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
       // RTO was spurious. Otherwise, inform congestion control.
       if (rto_count > 0 &&
             smallest_newly_acked > largest_sent_before_rto):
```
largest_sent_before_rto is still used here

nibanks replied:
As well as `rto_count`

siyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1183,14 +1142,14 @@ are detected lost.
 
 ~~~
-   OnRetransmissionTimeoutVerified(packet_number)
+   OnProbeTimeoutVerified(packet_number)
```
This looks like we will set the probetimeout verified even if this happens

    1, 2, 3 [lost], 4 [probe], 5 [probe]
    -> [ack 4]
    -> cwnd = mincwnd

in the first probe case it seems like the probe would be aggressive so it would be bad to drop the cwnd down.

even if it not aggressive right now, in the future I think we are going to tune timeouts constant per application (realtime vs interactive). This means that if we tune timeout constants too aggressively, it will result in having this reaction.

ianswett replied:
I'd suggest we keep a "you can not do the CWND reduction on the first or 2nd PTO" to keep behavior more similar to what QUIC has now.

nibanks commented on draft-ietf-quic-recovery.md:
```diff
@@ -361,8 +361,7 @@ increases loss detection delay.
 ack-based loss detection.  It uses a single timer which switches between
-a crypto retransmission timer, a Tail Loss Probe timer and Retransmission
-Timeout mechanisms.
+a crypto retransmission timer and a Probe timer.
```
You don't capitalize `Probe` in other places. It looks out of place here.

nibanks commented on draft-ietf-quic-recovery.md:
```diff
@@ -812,17 +785,11 @@ Pseudocode for SetLossDetectionTimer follows:
 
-    // RTO or TLP timer
-    // Calculate RTO duration
+    // PTO timer. Calculate PTO duration
```
Nit: I feel like having either `PTO timer` or `Calculate PTO duration` would be cleaner. Having both seems a bit overdone.

nibanks commented on draft-ietf-quic-recovery.md:
```diff
@@ -973,24 +936,20 @@ During recovery, the congestion window remains unchanged irrespective of new
+A PTO expiration is classified as spurious or valid when an ACK frame is
+received that newly acknowledges packets in flight, see {{pto-loss}}.  On a
+valid PTO, the congestion window MUST be reduced to the minimum congestion
+window and slow start is re-entered.
```
Just to confirm my understanding, this is a difference from TLP before, right? Before this change, if TLP triggered and you sent a probe packet that elicited an ACK for just that packet, you wouldn't go back into slow start, but now an ACK for only the first probe packet would trigger slow start. Correct?

ianswett replied:
Agreed, this is a potentially substantial change.

ianswett replied:
If you have a lot of tails(ie: HTTP traffic), then it doesn't take too many packet losses to end up hitting this.  TLP was targeted partially at cases like that, and I think you'll find changing it will produce measurable performance regressions in web traffic.  We can do something else in a follow-up, but as I said before, I'd prefer text saying SHOULD consider an acknowledgement of the first PTO as standard packet loss, not extreme congestion, and treat it as such.

janaiyengar replied:
I'm going to simplify this further. I think that having a verification is unnecessary; the behavior ends up being the same whether a PTO is verified or not. I can retain this notion of reducing the cwnd differently for the first 2 PTOs and then collapsing it to min_cwnd after that.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -328,8 +328,8 @@ The RECOMMENDED initial value for the packet reordering threshold
-detect spurious losses.  Implementers MAY use algorithms developed for TCP, such
-as TCP-NCR {{?RFC4653}}, to improve QUIC's reordering resilience.
+detect spurious losses.  Implementers MAY increase the packet threshold under
+such conditions.
```
This change seems unrelated to the rest of the change.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
+#### Computing PTO
 
-* PTO SHOULD be scheduled for max(1.5*SRTT+MaxAckDelay, kMinTLPTimeout)
+When the final ack-eliciting packet before quiescence is transmitted, the sender
```
Final is an optimization and we don't define quiescence
```suggestion
When an ack-eliciting packet is sent, the sender
```

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -577,21 +560,18 @@ kTimeThreshold:
-kMinTLPTimeout:
-: Minimum time in the future a tail loss probe timer may be set for.
-  The RECOMMENDED value is 10ms.
+kGranularity:
```
It feels odd to define this here, but use it three times up above.  I'd suggest discussing the idea that you're using clock granularity above where you first use it.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
 
-A PTO value of at least 1.5*SRTT ensures that the ACK is overdue.  The 1.5 is
-based on {{?TLP}}, but implementations MAY experiment with other constants.
+The PTO period is the amount of time that a sender ought to wait for an
```
```suggestion
The PTO period is the amount of time that a sender SHOULD wait for an
```

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-A PTO value of at least 1.5*SRTT ensures that the ACK is overdue.  The 1.5 is
-based on {{?TLP}}, but implementations MAY experiment with other constants.
+The PTO period is the amount of time that a sender ought to wait for an
+acknowledgement for a sent packet to be received.  This time period includes the
```
```suggestion
acknowledgement of a sent packet.  This time period includes the
```

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-based on {{?TLP}}, but implementations MAY experiment with other constants.
+The PTO period is the amount of time that a sender ought to wait for an
+acknowledgement for a sent packet to be received.  This time period includes the
+estimated network roundtrip-time (smoothed_rtt), the variance in the estimate
```
Using variance here is potentially confusing, since it's not true variance and it's 4*rttvar.

I'm having a hard time coming up with suggestions, but I'll think about it.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-timer be scheduled for a second time. When the TLP timer expires the second
-time, a second TLP packet is sent, and an RTO timer SHOULD be scheduled {{rto}}.
+There is no requirement on the clock granularity. If the PTO computation results
+in a value of zero, a sender MUST set the PTO value to kGranularity, to avoid
```
How about "The PTO must be set to at least kGranularity,"

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-used to send a probe into the network prior to establishing any packet loss,
-prior unacknowledged packets SHOULD NOT be marked as lost when a TLP timer
-expires.
+A PTO timer is set on an ack-eliciting tail packet.  A sender may not know that
```
Drop this first sentence and move the second sentence up with "When the final ack-eliciting packet"

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-conditions:
+When a PTO timer expires, the sender MUST send one ack-eliciting packet as a
+probe. A sender MAY send up to two ack-eliciting packets, to avoid an expensive
+consecutive PTO expiration due to packet loss.
```
```suggestion
consecutive PTO expiration due to a single packet loss.
```

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-* When an RTO timer expires, the RTO period is doubled.
+Probe packets sent on a PTO MUST be ack-eliciting.  A probe packet SHOULD carry
+new data when possible.  Implementers MAY use alternate strategies for
+determining the content of probe packets.  An implementation could use new data
```
From "An implementation could" on, this paragraph seems quite speculative, and the first sentence is redundant to the above setence "A probe packet SHOULD carry new data when possible."

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,88 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-this packet adds network load without establishing packet loss.
+If an ACK frame is received that newly acknowledges only those packets that were
+sent after a PTO timer expiration, all unacknowledged packets with lower packet
+numbers MUST be marked as lost.
```
This will happen naturally due to loss detection I believe.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -973,24 +936,20 @@ During recovery, the congestion window remains unchanged irrespective of new
+Probe packets MUST NOT be blocked by the congestion controller.  A sender MUST
+however count these packets as being additionally in flight, since these packets
+adds network load without establishing packet loss.  Note that sending probe
+packets might cause the sender's estimated bytes in flight to exceed the
```
```suggestion
packets might cause the sender's bytes in flight to exceed the
```

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -973,24 +936,20 @@ During recovery, the congestion window remains unchanged irrespective of new
+however count these packets as being additionally in flight, since these packets
+adds network load without establishing packet loss.  Note that sending probe
+packets might cause the sender's estimated bytes in flight to exceed the
+sender's congestion window until an acknowledgement is received that establishes
```
```suggestion
congestion window until an acknowledgement is received that establishes
```

ianswett replied:
Also, I think this would read better as "until packets are acknowledged or lost."

martinthomson commented on draft-ietf-quic-recovery.md:
```diff
@@ -297,7 +297,7 @@ If a packet is lost, the QUIC transport needs to recover from that loss, such
 frame.  For more information, see Section 13.2 of {{QUIC-TRANSPORT}}.
 
-## Ack-based Detection
+## Ack-based Detection {#ack-loss-detection}
```
```suggestion
## ACK-based Detection {#ack-loss-detection}
```

martinthomson commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,85 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
 
-* If RTO ({{rto}}) is earlier, schedule a TLP in its place. That is,
-  PTO SHOULD be scheduled for min(RTO, PTO).
+PTO = max(smoothed_rtt + 4*rttvar + max_ack_delay, kGranularity)
```
```suggestion
~~~
PTO = max(smoothed_rtt + 4*rttvar + max_ack_delay, kGranularity)
~~~
```

martinthomson commented on draft-ietf-quic-recovery.md:
```diff
@@ -844,16 +808,12 @@ Pseudocode for OnLossDetectionTimeout follows:
-       if (rto_count == 0)
-         largest_sent_before_rto = largest_sent_packet
+       // PTO.
+       if (pto_count == 0)
```
```suggestion
       if (pto_count == 0):
```

martinthomson commented:
Does this have an issue?

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,85 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
+data when new data is unavailable, when flow control does not permit new data to
+be sent, or to opportunistically reduce loss recovery delay.  Implementers MAY
+use alternate strategies for determining the content of probe packets, including
+sending new or retransmitted data based on the application's priorities.
```
Thanks, this is much better!

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,85 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-QUIC also diverges from TCP by including MaxAckDelay in the RTO period. Since
-QUIC corrects for this delay in its SRTT and RTTVAR computations, it is
-necessary to add this delay explicitly in the TLP and RTO computation.
+Delivery or loss of packets in flight is established when an acknowledgement is
```
Should this be "an ACK"?  We typically talk about an acknowledgement as referencing a single packet.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -415,100 +414,85 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-necessary to add this delay explicitly in the TLP and RTO computation.
+Delivery or loss of packets in flight is established when an acknowledgement is
+received that newly acknowledges one or more packets.  An endpoint might receive
+an acknowledgement after multiple consecutive PTO periods.
```
I feel like you're trying to say something more with this last sentence, but I'm not sure what?

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -802,7 +772,7 @@ Pseudocode for SetLossDetectionTimer follows:
       else:
         timeout = 2 * smoothed_rtt
-      timeout = max(timeout, kMinTLPTimeout)
+      timeout = max(timeout, granularity)
```
```suggestion
      timeout = max(timeout, kGranularity)
```

ianswett replied:
If you want to use granularity, then you need to define it above.  Otherwise, use kGranularity as suggested.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -812,17 +782,11 @@ Pseudocode for SetLossDetectionTimer follows:
-      tlp_timeout = max(1.5 * smoothed_rtt
-                        + max_ack_delay, kMinTLPTimeout)
-      timeout = min(tlp_timeout, timeout)
+    timeout = max(timeout, granularity)
```
```suggestion
    timeout = max(timeout, kGranularity)
```

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -722,14 +695,13 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
+      // If any packets sent prior to PTO were acked, then the
+      // PTO was spurious. Otherwise, inform congestion control.
+      if (pto_count > 0 &&
+            smallest_newly_acked > largest_sent_before_pto):
```
You could define a kNumCongestivePtoPackets here and add it here and then define it above to be the number of PTOs you will treat as standard packet loss, and not extreme congestion.

martinthomson replied:
What is "extreme congestion" ?

janaiyengar replied:
@martinthomson : @ianswett is trying to capture the difference in congestion response that the current loss recovery spec has for TLP vs. RTO. The idea is to say that for a couple of PTOs, there will be a smaller reduction in cwnd, and after that, the cwnd will be collapsed. I can write that up.

janaiyengar commented:
I've updated the description text and the PR quite a bit based on the comments. Specifically, I've removed the verification of a timeout, since the behavior with and without it was the same. I've also added a clause that allows for the cwnd to be collapsed to min cwnd only after 2 PTOs, to retain parity with TLP/RTO behavior.

I'll open an issue to discuss the issues this PR addresses. We should have discussion there.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -983,24 +926,18 @@ During recovery, the congestion window remains unchanged irrespective of new
-acknowledges no packets sent prior to the first retransmission timeout.  In this
-case, the congestion window MUST be reduced to the minimum congestion window and
-slow start is re-entered.
+If two consecutive PTOs have occurred (pto_count is at least 3), the network is
```
+1, I'm confused.  Also, this should be more about the number of PTOs that were lost than the number sent.   ie: I could send 3 PTOs, then the network returns to normal, no packets were lost, and the connection continues.

janaiyengar replied:
This is based on the number of timeouts that happened, not based on packets sent.  We want cwnd collapse at the equivalent of the first RTO's occurrence. This is the third timeout (2 TLP timeouts and then 1 RTO), which means pto_count is at least 3. Or am I doing this math wrong?

kazuho replied:
To me it seems that "pto_count is at least 3" is correct, however "If two consecutive PTOs have occurred" is incorrect and needs to be changed to "If three...".

pravb replied:
IIRC Ian was going to get data on whether we need 2 TLPs or just one is sufficient. FWIW we only do one TLP in Windows TCP. Yes at the least please fix as Kazuho says: "if three consecutive PTOs have occurred"

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -425,100 +425,79 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
+flight but an acknowledgement is not received within the expected period of
+time.  A PTO enables a connection to recover from loss of tail packets or acks.
+The PTO algorithm used in QUIC implements the reliability functions of Tail Loss
+Probe {{?TLP=I-D.dukkipati-tcpm-tcp-loss-probe}}, RTO {{?RFC5681}} and F-RTO
```
I think the function we're talking about is TLP, though I'm happy to add and cite RACK separately somewhere. I'll do that separately.

pravb replied:
TLP is an expired draft in 2013 and RACK is standards track and active. Linux, FreeBSD and Windows are all implementing the RACK draft and TLP as specified in the RACK draft. I would just cite the RACK draft. There is a chance it may got into last call within the next two IETFs.

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -587,21 +562,18 @@ kTimeThreshold:
 kDelayedAckTimeout:
 : The length of the peer's delayed ack timer. The RECOMMENDED value is 25ms.
 
 kInitialRtt:
```
Would you mind opening a separate issue for this? I'd argue for 100ms, but I don't want to do it in this PR. FWIW, I think the cost of an early timeout is not high, so having this be aggressive seems fine to me.

ianswett replied:
Agreed, that's a separable issue, which was previously raised and then I believe closed.

pravb replied:
Ok I will open a separate issue. This will cause unnecessary timeouts for CI and 0-RTT data on many Internet paths. And pretty much always on satellite links. If we want to pick a magic number then let's pick 200 msec or 300 msec (the historical minrto values for Linux and Windows respectively).

mikkelfj replied:
Here are 10 ping packets from Denmark to China

10 packets transmitted, 10 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 211.218/350.768/593.271/149.383 ms

mikkelfj replied:
And here from Denmark to New York, which is basically a direct sea cable link.

10 packets transmitted, 10 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 103.982/104.505/105.343/0.475 ms

mikkelfj replied:
And Denmark New Zealand

10 packets transmitted, 10 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 324.042/454.764/672.469/122.684 ms

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -1157,6 +1095,9 @@ window.
        congestion_window = max(congestion_window, kMinimumWindow)
        ssthresh = congestion_window
+       // Collapse congestion window if persistent congestion
+       if (pto_count > 2):
```
2 is a magic number.  Previously we said you SHOULD have 2 TLPs, but you could have fewer.  I think it makes sense to define this as a constant and explain its reason for existing there?

pravb replied:
I commented on the thread above this. Yes we should settle on one or two TLPs. I personally prefer one but unfortunately I don't have data to say if two is better since Windows TCP has been doing just one TLP.

janaiyengar replied:
@siyengar : In this case, by the time 5 is acked, 3 and 4 should be marked as lost by the time-threshold loss detection mechanism, since they're going to be > 1.125 x SRTT in the past.

pravb commented on draft-ietf-quic-recovery.md:
```diff
@@ -425,100 +425,80 @@ from when the first Initial was sent to when a Retry or a Version Negotiation
-timer be scheduled for a second time. When the TLP timer expires the second
-time, a second TLP packet is sent, and an RTO timer SHOULD be scheduled {{rto}}.
+~~~
+PTO = max(smoothed_rtt + 4*rttvar + max_ack_delay, kGranularity)
```
Sorry just noticed another difference. If peer advertises max_ack_delay as 0 (which seems to be legal), then first PTO becomes Srtt + 4 * RttVar, whereas in TCP the first PTO is 2* Srtt. If variance is low this may be too aggressive? Should we also do PTO = max (PTO, 2 *Srtt) ?

janaiyengar replied:
I'll push back against this. I like the simplicity of having a single computation that includes rttvar. What is the rationale in TCP for using 2xSRTT when RTTVAR is available? If variance is low, that still seems like the right thing to do. Why inflate the PTO when you don't need to?

FWIW, the first PTO is akin to the first TLP. The second PTO would be backed off, so you will automatically get 2 x (SRTT + 4xRTTVAR).

pravb replied:
Quoting RACK draft:

Aiming for a PTO value of 2*SRTT allows a sender to wait long enough
   to know that an ACK is overdue.  Under normal circumstances, i.e. no
   losses, an ACK typically arrives in one SRTT.  But choosing PTO to be
   exactly an SRTT is likely to generate spurious probes given that
   network delay variance and even end-system timings can easily push an
   ACK to be above an SRTT.  We chose PTO to be the next integral
   multiple of SRTT.

   Similarly, network delay variations and end-system processing
   latencies and timer granularities can easily delay ACKs beyond
   2*SRTT, so senders SHOULD add at least 2ms to a computed PTO value
   (and MAY add more if the sending host OS timer granularity is more
   coarse than 1ms).

janaiyengar replied:
@pravb: Thanks for the context -- I should have just looked at the RACK draft. So, to be clear, the RACK draft does not use this 2ms when WCDelAckT is used, that is, when it accounts for the worst case delayed ack timer. This is what we expect max_ack_delay to do for us, but as you note, a receiver might send 0 here. 

I think we should put the onus of this onto the receiver.  A receiver should send the worst case ack delay in max_ack_delay. A receiver that doesn't do that shoots itself in the foot by causing spurious retransmissions at the sender. We may want to add some strong wording about how a receiver should choose max ack delay, I've filed #2186 for that.

Sound reasonable?

pravb replied:
That's reasonable. So this is how we'd justify the deviation from TCP:
1. Include RttVar component in the computation for in network variability. 
2. Require the receiver to include the perturbation for system processing delays. So 2186 needs to add a SHOULD for adding a perturbation value on top of the actual max ack delay. I'll comment there.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -1065,6 +1004,16 @@ kLossReductionFactor:
+  Loss Probe (TLP) in TCP {{TLP}} {{RACK}}.  Once the number of consecutive PTOs
+  reaches this threshold - that is, persistent congestion is established - the
+  sender responds by collapsing its congestion window, similar to a
+  Retransmission Timeout (RTO) in TCP {{RFC5681}}.  The RECOMMENDED value is 3,
```
The pseudocode below decreases CWND to kMinimumWindow, which is 2 packets.  I think sticking with that makes sense, at least for now.  It also requires less text to justify.

janaiyengar replied:
The recommendation is 3 for kPersistentCongestion, not for the value of cwnd. I'll change the text a bit to clarify.

marten-seemann commented:
I think I found the answer to my question:
Since in #1974 we make time-threshold based loss recovery mandatory (before that you could choose between packet-threshold and time-threshold), we don't need the RTO verified anymore (see #2168) anyway. This is independent of this PR, this PR just happens to fix multiple issues at the same time.

Sorry for the noise.

janaiyengar commented:
There's exactly one question that's pending to be resolved now in this PR: @pravb's question about using 2SRTT for initial PTO. I'm happy to open a new issue for that specific question if Praveen thinks we should continue that discussion, but I'd like to merge this soon, as (i) I'm hearing pushback on a detail or two but not the overall PR, which we can handle in separate issues, and (ii) this PR is fast approaching github's limit on comments (hence the unicorn tag).
